sklearn/tree/_criterion.pyx

```diff
@@ -574,20 +575,26 @@ cdef class ClassificationCriterion(Criterion):
             memcpy(dest, &self.sum_total[k, 0], self.n_classes[k] * sizeof(float64_t))
             dest += self.max_n_classes
 
-    cdef void clip_node_value(self, float64_t * dest, float64_t lower_bound, float64_t upper_bound) noexcept nogil:
-        """Clip the value in dest between lower_bound and upper_bound for monotonic constraints.
-
+    cdef void clip_node_value(
+        self, float64_t * dest, float64_t lower_bound, float64_t upper_bound
+    ) noexcept nogil:
+        """Clip the values in dest such that predicted probabilities stay between
+        `lower_bound` and `upper_bound` when monotonic constraints are enforced.
         Note that monotonicity constraints are only supported for:
         - single-output trees and
         - binary classifications.
         """
-        if dest[0] < lower_bound:
-            dest[0] = lower_bound
-        elif dest[0] > upper_bound:
-            dest[0] = upper_bound
-
-        # Class proportions for binary classification must sum to 1.
-        dest[1] = 1 - dest[0]
+        cdef:
+            float64_t total_weighted_count = dest[0] + dest[1]
+            float64_t scaled_lower_bound = (<float32_t>lower_bound) * total_weighted_count
+            float64_t scaled_upper_bound = (<float32_t>upper_bound) * total_weighted_count
```

Comment:
```suggestion
            float64_t scaled_lower_bound = <float32_t>(lower_bound * total_weighted_count)
            float64_t scaled_upper_bound = <float32_t>(upper_bound * total_weighted_count)
```
To void early loss of precision.
Why do we need to cast to float32 at all?

Reply:
> To void early loss of precision.
> Why do we need to cast to float32 at all?

The precision loss for `lower_bound` and `upper_bound` (and not for their scaled counterpart) is by design.

The subtlety here is that for leaves affected by clipping (say constrained by `lower_bound`), the predicted probabilities (without casting) would be `(lower_bound * total_weighted_count) / total_weighted_count`, which varies with `total_weighted_count` close to `lower_bound`. 
So if two leaves end up having the same `lower_bound` desired prediction, but two different `total_weighted_count`, their predictions might end up being (slightly) different and then (slightly) breaking the monotonicity constraint.

By casting `lower_bound` to `float32` we make `((<float32_t>lower_bound) * total_weighted_count) / total_weighted_count` effectively constant for a pretty large set of `total_weighted_count` values, most interestingly integers < 2**30. The reason is that:
- the multiplication by `total_weighted_count` in this subset of values can be done *without rounding*
- the division by `total_weighted_count` (the *same number*) will be *exact* thus requiring no rounding either and equal to the initial `<float32_t>lower_bound`.

Please let me know if any of this is not clear enough !

Reply:
Thanks for the suggestions. Adding an explanatory comment sounds like a good idea, not sure how short I can make it though...

Regarding the precision loss on `lower_bound` and `upper_bound`, we want the sequence of multiplication and division by the same number to be exactly neutral, which requires more loss of precision than a single bit.
As you can see that 
```python
def check_nextafter(a, n): return (nextafter(a, 1) * n) / n == nextafter(a, 1)
check_nextafter(0.8793444460978181, 700)
>>> False
```
whereas 
```python
def check_castfloat32(a, n): return (np.float32(a) * n) / n == np.float32(a)
```
requires n > 2**29 for returning `False`, which seems good enough.

Reply:
That's not a fair comparison and not the one we are interested in. We want `scaled_lower_bound / total_weighted_count >= lower_bound`, but as close as possible (that's where float32 sucks, in the end, it might not influence the estimator in a noticable way).

```python 
from math import nextafter
import numpy as np

def check_nextafter(a, n):
     """a between 0 and 1, n any number"""
     return (nextafter(a, 1) * n) / n >= a

sum([not check_nextafter(0.8793444460978181, n) for n in range(1, 1000_000)])  # 0
sum([not check_nextafter(0.8793444460978181, n) for n in np.logspace(0, 100,  1_000_000)])  # 0
```

Reply:
@glemaitre @betatim @jjerphan Could you step in?

As long as we use a tiny larger lower bound, monotonicity will always be fulfilled. `nextafter` will also always produce the same number.
Given that the "mid value" is a bit arbitrary, we are discussing around that arbitrariness.

I still hold to my proposal and could be convinced by a (otherwise) failing test.

Reply:
> As long as we use a tiny larger lower bound, monotonicity will always be fulfilled.

Actually no (I explained why in my last comment). Besides, I just ran tests with your suggestion and they do fail. It's still a pretty rare occurence but that's precisely the case that the `float32` casting is addressing.

> `nextafter` will also always produce the same number.

Sure, but not `(nextafter * n) / n`, which is what matters when we compare predictions.

Reply:
> I still hold to my proposal and could be convinced by a (otherwise) failing test.

Specifically, `test_monotonic_constraints_classifications` with `107` as `global_random_seed`:
- fails with your proposal (on a monotonicity test)
- passes with the current code.

Reply:
Interestingly, the same test passes with two rounds of `nextafter` ! But then with seed `644` the test fails again. 
If I add third round of `nextafter` the test passes, but then with seed `1559` it fails again...

Reply:
> @glemaitre @betatim @jjerphan Could you step in?

I would like to get some time to catch up on the discussions.

Reply:
If we go with casting, we should use float32 for the bounds right from the start.

Reply:
I ran a few tests with the `nextafter` proposal while introducing a non-trivial `sample_weight` to the training and it works much better than casting to `float32` (still failing between 2% and 3% of the time, compared to more than 80% for `float32` casting).

Reply:
OK so it seems that this is a better solution and we should be leaning towards this solution

Regarding the remaining failures, could we avoid them by tacking `nextafter` with a larger `steps`?

Reply:
Which assert in which test does fail?
I don't now of a "multi_nextafter", but we could just use `lower_bound * (1 + x * np.finfo(float).eps)` and `lower_bound * (1 - x * np.finfo(float).epsneg)` for some `x > 1`.

Reply:
> Which assert in which test does fail?

In `test_monotonic_constraints_classifications`:
```python
assert np.all(est.predict_proba(X_test_0incr)[:, 1] >= proba_test[:, 1])
```

Reply:
> the `nextafter` from python as a `steps` parameter.

Indeed, this would be useful, although it was only [added](https://docs.python.org/3/library/math.html#math.nextafter) in Python 3.12...

Reply:
Arff my bad. I did not check the full documentation. Python 3.12 is quite ahead.